Machine ID: Terraform Cloud joining

timothyb89 opened the pull request:
This adds a new `terraform` join method, which uses Terraform Cloud's Workload Identity OIDC provider to provide delegated joining support for Machine ID bots. As the Terraform provider now embeds tbot, this means - with some light modifications - that the Terraform provider can now join using this join method, and without any secrets.

Short instruction doc: https://gist.github.com/timothyb89/5bb905e8c7065ae0601e83f816af8056

changelog: Add new `terraform` join method to allow secretless authentication for Terraform Cloud jobs in the Teleport Terraform provider

---

Separate docs PR (WIP): https://github.com/gravitational/teleport/pull/45806

timothyb89 committed:
Machine ID: Terraform Cloud joining

This adds a new `terraform` join method, which uses Terraform Cloud's
Workload Identity OIDC provider to provide delegated joining support
for Machine ID bots. As the Terraform provider now embeds tbot, this
means - with some light modifications - that the Terraform provider
can now join using this join method, and without any secrets.

strideynet commented on api/proto/teleport/legacy/types/types.proto:
```diff
@@ -1579,6 +1581,43 @@ message ProvisionTokenSpecV2GCP {
+
+  // Audience is the JWT audience as configured in the
+  // TFC_WORKLOAD_IDENTITY_AUDIENCE(_$TAG) variable in Terraform Cloud.
+  string Audience = 2 [(gogoproto.jsontag) = "audience,omitempty"];
```
Could we have it so that we always just expect the audience to equal the name of the Teleport cluster ? That sounds more semantically correct to me.

timothyb89 replied:
We could, and I think that's the only answer we could bake in that makes sense. It might make the config/docs UX a bit weird (without a guided flow of some sort at least) but is otherwise fine.

The more confusing parameter is the `$TAG`, I suppose. For docs purposes we can just recommend `TFC_WORKLOAD_IDENTITY_AUDIENCE_TELEPORT=<cluster name>`.

Do you think that should be the default value if unset, or should we just always use the cluster name? Is there any sane usecase for allowing user override?

strideynet commented on api/types/provisioning.go:
```diff
@@ -817,3 +832,28 @@ func (a *ProvisionTokenSpecV2TPM) validate() error {
+
+	for i, allowRule := range a.Allow {
+		orgSet := allowRule.OrganizationID != "" || allowRule.OrganizationName != ""
+		projectSet := allowRule.ProjectID != "" || allowRule.ProjectName != ""
```
Are project names globally unique across TfCloud ? We should ensure we require at least one globally unique property here.

timothyb89 replied:
Organization names are, at least. I think we could require that at least organization (name or ID) is set, then any combination of project/workspace?

strideynet replied:
👍 I think I've done similar on another one of the join methods (always required org id/org name). Without it, it's super easy to make a join token that would allow a bad actor to create their own org, create a project w/ the same name, and access your cluster.

strideynet commented on lib/terraform/token_source.go:
```diff
@@ -0,0 +1,65 @@
+func (its *IDTokenSource) GetIDToken() (string, error) {
+	name := "TFC_WORKLOAD_IDENTITY_TOKEN"
+	if its.audienceTag != "" {
+		name = fmt.Sprintf("TFC_WORKLOAD_IDENTITY_TOKEN_%s", strings.ToUpper(its.audienceTag))
```
Wow - had no clue they put the audience into the name of the env var - that's super awkward!

timothyb89 replied:
Yep, it's a weird one. Well, not quite in the _name_ of the var - the audience tag is an optional parameter in the var name. The audience is the var value.

The rule is more or less: whenever TFC sees a var matching `TFC_WORKLOAD_IDENTITY_AUDIENCE(?P<tag>_\w+)?` during a run, it uses the value as the audience, then writes the JWT into `$TFC_WORKLOAD_IDENTITY_TOKEN$tag`.

The tag is meant to let you issue more than one token, since otherwise you could only define one audience. More docs here: https://developer.hashicorp.com/terraform/enterprise/workspaces/dynamic-provider-credentials/specifying-multiple-configurations

---

(Naturally, none of this applies to their own providers which use some slightly different magic to write the same content into different env vars 🤷)

strideynet commented on lib/auth/join_terraform.go:
```diff
@@ -0,0 +1,103 @@
+		return nil, trace.BadParameter("terraform join method only supports ProvisionTokenV2, '%T' was provided", pt)
+	}
+
+	if modules.GetModules().BuildType() != modules.BuildEnterprise {
```
I'm 50-50 on making this enterprise only. We should double check with Product.

timothyb89 committed:
Address some review feedback, add first batch of tests

This addresses a few feedback items, defaulting the audience to the
Teleport cluster name and requiring allow rules specify an
organization + at least one other parameter.

Also, adds a first batch of tests.

timothyb89 committed:
Update autogenerated CRD docs

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/terraform-cloud-joining

timothyb89 committed:
Add tests for RegisterUsingToken and provisioning token init

timothyb89 committed:
Fix imports

timothyb89 committed:
Fix CRD docs

timothyb89 committed:
Update tfschema

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/terraform-cloud-joining

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/terraform-cloud-joining

timothyb89 committed:
Update tf docs

timothyb89 committed:
Update docstring to make it clear that `Audience` is optional

timothyb89 committed:
Update manifests

timothyb89 committed:
Update CRD docs

marcoandredinis commented on api/proto/teleport/legacy/types/types.proto:
```diff
@@ -1580,6 +1582,44 @@ message ProvisionTokenSpecV2GCP {
+message ProvisionTokenSpecV2Terraform {
+  // Rule is a set of properties the Terraform-issued token might have to be
+  // allowed to use this ProvisionToken.
+  message Rule {
```
Is every field required? Or is there any optional one?

timothyb89 replied:
Technically they are all optional with some verification rules (some org name/id + at least one other parameter). I'll make sure the docstring here mentions it.

marcoandredinis commented on api/proto/teleport/legacy/types/types.proto:
```diff
@@ -1580,6 +1582,44 @@ message ProvisionTokenSpecV2GCP {
+    string WorkspaceName = 6 [(gogoproto.jsontag) = "workspace_name,omitempty"];
+
+    // RunPhase is the phase of the run the token was issued for, e.g. `plan` or `apply`
+    string RunPhase = 7 [(gogoproto.jsontag) = "run_phase,omitempty"];
```
Should this be a list? Eg, allowing for the `plan` and `apply` phase

timothyb89 replied:
It can only be one value at a time, and users can specify multiple allowed values using several `allow` rules if they want. Truthfully I'm not even sure whether or not there's a real use case for restricting the run phase, but I figured maybe someone could get creative with it.

An enum would probably be nicer than a string, but annoyingly Terraform doesn't define/document an exhaustive list of values and only provide "plan" and "apply" as examples of possible values :/

marcoandredinis commented on api/proto/teleport/legacy/types/types.proto:
```diff
@@ -1580,6 +1582,44 @@ message ProvisionTokenSpecV2GCP {
+  // allow rule to use this token.
+  repeated Rule Allow = 1 [(gogoproto.jsontag) = "allow,omitempty"];
+
+  // Audience is the JWT audience as configured in the
+  // TFC_WORKLOAD_IDENTITY_AUDIENCE(_$TAG) variable in Terraform Cloud. If
+  // unset, defaults to the Teleport cluster name.
```
Can we please add an example of what the env var looks like?

marcoandredinis commented on lib/terraform/token_source.go:
```diff
@@ -0,0 +1,65 @@
+	getEnv envGetter
+}
+
+func (its *IDTokenSource) GetIDToken() (string, error) {
```
💅 Missing godoc

marcoandredinis reviewed (approved):
I would change the joining method name from Terraform into TerraformCloud.
I was confused at first when reading the code, because I assumed it could be somehow used for every terraform deployment (not sure how that would work).

Other than that, looks good 👍

timothyb89 commented:
> I would change the joining method name from Terraform into TerraformCloud. I was confused at first when reading the code, because I assumed it could be somehow used for every terraform deployment (not sure how that would work).

Hmm, yeah, I'm feeling the same way. We were initially unsure if it would be TFC exclusive but it does appear to be now. I'll work on a rename.

timothyb89 committed:
Code review feedback; docstring fixes

timothyb89 committed:
Rename lib/terraform to lib/terraformcloud

timothyb89 committed:
Rename terraform -> terraform_cloud

Renames the new join method to terraform_cloud to make it clear that
it won't work for TF running elsewhere (where other join methods
should be used)

timothyb89 committed:
Fix tests and docs lints

mgriffin13 commented:
> > I would change the joining method name from Terraform into TerraformCloud. I was confused at first when reading the code, because I assumed it could be somehow used for every terraform deployment (not sure how that would work).
> 
> Hmm, yeah, I'm feeling the same way. We were initially unsure if it would be TFC exclusive but it does appear to be now. I'll work on a rename.

FWIW, I'd imagine this same solution would be used for Terraform Enterprise (the self-hosted version of TFC). Not sure that's enough to preclude a name change, but it might be worth mentioning in the docs at least.

https://developer.hashicorp.com/terraform/enterprise/workspaces/dynamic-provider-credentials/workload-identity-tokens

tigrato commented on api/proto/teleport/legacy/types/types.proto:
```diff
@@ -1296,6 +1296,8 @@ message ProvisionTokenSpecV2 {
   ProvisionTokenSpecV2Spacelift Spacelift = 14 [(gogoproto.jsontag) = "spacelift,omitempty"];
   // TPM allows the configuration of options specific to the "tpm" join method.
   ProvisionTokenSpecV2TPM TPM = 15 [(gogoproto.jsontag) = "tpm,omitempty"];
+  // Terraform allows the configuration of options specific to the "terraform_cloud" join method.
```
```suggestion
  // TerraformCloud allows the configuration of options specific to the "terraform_cloud" join method.
```

timothyb89 commented:
> FWIW, I'd imagine this same solution would be used for Terraform Enterprise (the self-hosted version of TFC). Not sure that's enough to preclude a name change, but it might be worth mentioning in the docs at least.
> 
> https://developer.hashicorp.com/terraform/enterprise/workspaces/dynamic-provider-credentials/workload-identity-tokens

The docs aren't super clear here so I think we'd need to work with a customer running Terraform Enterprise to test it, but this _should_ be a reasonably simple addition we can make in a followup. It looks like the issuer for Enterprise will just be the self hosted cluster, so we'll just make that URL configurable on our end (as part of the provision token).

In terms of naming, I think it'll be less weird to use a `terraform_cloud`-type token with an overridden issuer than to keep it named just `terraform`, which I think already creates confusion between non-cloud Terraform users and our own Terraform provider.

timothyb89 committed:
Update api/proto/teleport/legacy/types/types.proto

Co-authored-by: Tiago Silva <tiago.silva@goteleport.com>

timothyb89 committed:
Update proto artifacts

mgriffin13 commented:
> > FWIW, I'd imagine this same solution would be used for Terraform Enterprise (the self-hosted version of TFC). Not sure that's enough to preclude a name change, but it might be worth mentioning in the docs at least.
> > https://developer.hashicorp.com/terraform/enterprise/workspaces/dynamic-provider-credentials/workload-identity-tokens
> 
> The docs aren't super clear here so I think we'd need to work with a customer running Terraform Enterprise to test it, but this _should_ be a reasonably simple addition we can make in a followup. It looks like the issuer for Enterprise will just be the self hosted cluster, so we'll just make that URL configurable on our end (as part of the provision token).
> 
> In terms of naming, I think it'll be less weird to use a `terraform_cloud`-type token with an overridden issuer than to keep it named just `terraform`, which I think already creates confusion between non-cloud Terraform users and our own Terraform provider.

Thanks. I'm excited for this, because this is blocking my upgrade to Teleport v16+ after the change that required MFA for every local user, and I'm using Terraform Enterprise, which is why I'm curious about compatibility.

timothyb89 commented:
> Thanks. I'm excited for this, because this is blocking my upgrade to Teleport v16+ after the change that required MFA for every local user, and I'm using Terraform Enterprise, which is why I'm curious about compatibility.

@mgriffin13, if you'd be up for testing out early builds, feel free to reach out to me on our community Slack! I'd like to follow up quickly with support for Terraform Enterprise but don't have a way of testing it myself at the moment. If things go well, I think we should be able to get everything into v16.3.

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/terraform-cloud-joining
